### PR TITLE
Enhances mob AI by letting them advance toward you if they can see you but can't shoot

### PR DESCRIPTION
**Pull request by nikothedude**
## About The Pull Request
Mobs will now fire a invisible hitscan tracer if they can see you, and if it hits something that isn't their target, they will advance a few tiles forward, forcing them to pathfind.

Basic local testing shows no runtimes, proper functionality, etc. 

## Changelog
:cl:
balance: Mobs will now pathfind if they can't hit you.
/:cl:

### CODE REVIEW

File: code/__DEFINES/components.dm [hunk 1]
```diff
@@ -76,6 +80,8 @@
 #define COMSIG_MOB_LOGIN "mob_login"							//from mob/Login()
 #define COMSIG_MOB_DEATH "mob_death"							//from mob/death()
 
+#define COMSIG_TRACE_IMPACT "trace_impact"
```
Comment (Trilbyspaceclone) [inline code/__DEFINES/components.dm:83]:
```suggestion
#define COMSIG_TRACE_IMPACT "trace_impact" //Soj edit
```

**Comment by Trilbyspaceclone** [comment]
Please make sure to slap in `SoJ edit`
to some of the more major files so porting is slightly easyer

**Comment by nikothedude** [comment]
Mostly untested, but it works. 

I can't test this anymore, I leave in 15 minutes for maine, so I've gotta finish this up.

If anyone wants to finish this while I'm gone, AKA just polish it up, I'd be grateful.

**Comment by nikothedude** [comment]
Oh yeah also you'd need to fix the signals. That's a thing.

File: code/modules/mob/living/carbon/superior_animal/superior_animal.dm [hunk 2]
```diff
@@ -196,7 +202,7 @@
 	var/ranged = FALSE  //Do we have a range based attack?
 	var/rapid = FALSE   //Do we shoot in groups?
 	var/rapid_fire_shooting_amount = 3 //By default will rapid fire in 3 shots per.
-	var/projectiletype  //What are we shooting?
+	var/obj/item/projectile/projectiletype  //What are we shooting?
```
Comment (Trilbyspaceclone) [inline code/modules/mob/living/carbon/superior_animal/superior_animal.dm:205]: Why are you doing this?
Dosnt this limit to only being projectiles being exspected?

**Comment by nikothedude** [comment]
finally getting back to this

File: code/modules/mob/living/carbon/superior_animal/superior_animal.dm (hunk 2 above)
Comment (nikothedude) [reply code/modules/mob/living/carbon/superior_animal/superior_animal.dm:205]: i mean, we dont have aby mob that fires something that isnt a projectile

Comment (nikothedude) [reply code/modules/mob/living/carbon/superior_animal/superior_animal.dm:205]: plus i just was told that assigning a variable of a different type to a typed variable doesnt cause bad things as long as you dont try to access typed vars

**Comment by nikothedude** [comment]
TODO: convert can_see into this same type of check because this is a really really really good check

**Pull request merged**